src/bt-core-io.c

```diff
@@ -465,27 +468,78 @@ energy_info_cb(bt_activity_energy_info *energy_info ATTRIBS(UNUSED))
   /* nothing to do */
 }
 
+struct alarm_state {
+  alarm_cb cb;
+  void* data;
+};
+
+static enum ioresult
+alarm_event_in(int fd, void* data)
+{
+  uint64_t num_expirations;
+  ssize_t res;
+
+  res = TEMP_FAILURE_RETRY(read(fd, &num_expirations, sizeof(num_expirations)));
+  if (res < 0) {
+    ALOGE_ERRNO("read");
+    return IO_ABORT;
+  }
+
+  remove_timer(fd);
+  free(data);
+  return IO_OK;
+}
+
+static enum ioresult
+alarm_event(int fd, uint32_t events, void* data)
+{
+  enum ioresult res;
+
+  if (events & EPOLLIN) {
+    res = alarm_event_in(fd, data);
+  } else {
+    ALOGW("unsupported event mask: %u", events);
+    res = IO_OK;
+  }
+  return res;
+}
+
 static bool
-set_wake_alarm(uint64_t delay_millis ATTRIBS(UNUSED),
-               bool should_wake ATTRIBS(UNUSED),
-               alarm_cb cb ATTRIBS(UNUSED),
-               void *data ATTRIBS(UNUSED))
+set_wake_alarm_os_callout(uint64_t delay_millis, bool should_wake,
+                          alarm_cb cb, void* data)
 {
-  /* FIXME: need to be implemented in later patches */
+  struct alarm_state* alarm_state;
+  int clockid;
+
+  alarm_state = malloc(sizeof(*alarm_state));
+  if (!alarm_state)
```

Comment:
The pattern is slightly problematic. POSIX permits |{c,m,re}alloc| to return NULL if the allocation size is zero. In case that 'sizeof(*alarm_state)' is 0, the code might fail for no reason. I know that the size won't be 0 in practice, but I'd like to have the code cover all cases.

The correct pattern is

errno = 0;
mem = malloc();
if (errno) {
  return false;
}

errno will be ENOMEM if allocation failed and 0 in all other cases.